h/services/job_queue.py

```diff
@@ -0,0 +1,193 @@
+
+    def add_sync_annotation_jobs(self, annotation_ids, tag, scheduled_at=None):
+        """Add a list of annotations to the queue to be synced to Elasticsearch."""
+        self._db.add_all(
+            Job(
+                tag=tag,
+                scheduled_at=scheduled_at,
+                kwargs={"annotation_id": annotation_id},
+            )
+            for annotation_id in annotation_ids
+        )
+
+    def sync_annotations(self):
+        """
+        Synchronize annotations from Postgres to Elasticsearch.
+
+        This method is meant to be run periodically. It's called by a Celery
+        task that's scheduled periodically by h-periodic.
+
+        Each time this method runs it considers a fixed number of sync
+        annotation jobs from the job queue. For each job, if the annotation is
+        already present and up to date in Elasticsearch, then it removes the
+        job from the queue. If the annotation is missing from Elasticsearch or
+        out of date in Elasticsearch then the method re-syncs the annotation
+        into Elasticsearch, and leaves the job on the queue to be re-checked
+        and removed the next time the method runs.
+
+        If the DB somehow contains an annotation that always fails to index
+        into Elasticsearch, then a job for that annotation will never be
+        removed from the queue! As far as I know it isn't possible for the DB
+        to contain such an annotation.
+        """
+        jobs = self._get_jobs_from_queue()
+
+        if not jobs:
+            return
+
+        annotation_ids = [job.kwargs["annotation_id"] for job in jobs]
+        annotations_from_db = self._get_annotations_from_db(annotation_ids)
+        annotations_from_es = self._get_annotations_from_es(annotation_ids)
+
+        missing_from_db = []
+        missing_from_es = []
+        outdated_in_es = []
+        up_to_date_in_es = []
+
+        for job in jobs:
+            annotation_id = job.kwargs["annotation_id"]
+            annotation_from_db = annotations_from_db.get(annotation_id)
+            annotation_from_es = annotations_from_es.get(annotation_id)
+
+            if not annotation_from_db:
+                missing_from_db.append(job)
+            elif not annotation_from_es:
+                missing_from_es.append(job)
+            elif annotation_from_es["updated"] != annotation_from_db.updated:
+                outdated_in_es.append(job)
+            else:
+                up_to_date_in_es.append(job)
+
+        if missing_from_db:
+            logger.info(
+                f"Deleting {len(missing_from_db)} sync annotation jobs because their annotations have been deleted from the DB"
+            )
+        if up_to_date_in_es:
+            logger.info(
+                f"Deleting {len(up_to_date_in_es)} successfully synced annotations from job queue"
+            )
+
+        for job in missing_from_db + up_to_date_in_es:
+            self._db.delete(job)
```

Comment:
I thought this _was_ batched into one query by SQLAlchemy, but I can check that. I think the call to `delete()` just marks the ORM object as deleted in memory, and then when the SQLAlchemy session is flushed or committed SQLAlchemy batches all the in-memory ORM changes into as few SQL queries as possible. I'll see if I can double check what SQL queries it's actually emitting though.

The SO answer is about something else, I think. He's asking about how to reduce a single `SELECT` followed by a single `DELETE` (of the same rows that were selected) into just a `DELETE`. (Rather than how to reduce multiple `DELETE`s into one `DELETE`, which I _think_ SQLAlchemy just does.)

Reply:
Fair enough!

Reply:
I've tested this and SQLAlchemy does batch them for us